### ISSUE TITLE
Rethink error-handling

### ISSUE THREAD

**Issue by Pathoschild** [Pathoschild/FluentHttpClient#38]
The client throws exceptions by default for non-success responses. Should we rethink the default error-handling logic?

**Comment by Pathoschild** [comment]
Split from #36. Related discussion from that ticket:

* @Jericho said:
  >Originally I was strongly in favor of removing the default error handling for at least two reasons:
  >
  >- when I started using the FluentHttpClient package I was not aware that this behavior was overidable(I guess I didn't pay attention to the 'Error Handling' section in the ReadMe!)
  >- I was trying to write my own retry logic to handle HTTP 429 "Too many requests" but the default error handling was interfering because it would throw an exception before I could handle the response. Therefore my code was full of `try ... catch`.
  >
  >However, since we recently added built-in retry logic to FluentHttpClient which automatically retries requests before the error handler processes the response, I no longer have such a strong opinion. I'm fine with keeping the default error handler, especially since I now know that I can remove it and replace it with my own error handling if desired.

* @Pathoschild said:
  >The current error-handling behaviour can be counterintuitive. For example, consider the following code:
  >```cs
  >IResponse response = await client.GetAsync(…);
  >if (response.StatusCode != HttpStatusCode.OK)
  >    …
  >```
  >
  >If the response code isn't `HttpStatusCode.OK`, the `if` check will never be reached since an exception will be thrown instead. The only way around that would be to remove the `ErrorFilter` from the client before making the request. I'm not sure what the best way to handle that would be yet, though.

* @Jericho said:
  > **Re Error handling**: what you just described is exactly what I experienced when I started using FluentHttpClient and I had to change all my code which was checking the status code of the response with try ... catch. I didn't realize that the error handling filter could be removed until you told me. People who are transitioning from using the .net HttpClient to this client are used to checking the status code therefore I think it would be preferable not to register the error handling filter. You could leave the class in the project which would allow developers to use it if they like it.

**Comment by Pathoschild** [comment]
## How other clients do it
Here's how some of the other C# API clients handle errors. For each one, I added an example of how you'd fetch a deserialised model but throw a relevant exception with the response data on error.

### `HttpClient`
The built-in client doesn't automatically handle errors, but it provides a `response.IsSuccessStatusCode` property.

```cs
HttpResponseMessage response = await client.GetAsync("…");
if (!response.IsSuccessStatusCode)
{
   string content = response.Content.ReadAsStringAsync();
   throw new Exception("Server responded {response.StatusCode}: {content}");
}
return response.Content.ReadAsAsync<T>();
```

### RestSharp
[RestSharp](https://github.com/restsharp/RestSharp) doesn't automatically handle errors, and only treats network transport problems as errors (e.g. error 404 is considered a success). Example derived from [StackOverflow](https://stackoverflow.com/questions/31516501/how-to-idiomatically-handle-http-error-codes-when-using-restsharp).

```cs
var response = client.Execute<T>(request);
bool isSuccessCode = (int)response.StatusCode >= 200 && (int)response.StatusCode <= 399;
if (!isSuccessCode || response.ErrorException != null)
    throw new Exception($"Server responded {response.StatusCode}: {response.Content}", response.ErrorException);
return response.Data;
```

### Flurl
[Flurl](https://github.com/tmenier/Flurl) throws exceptions on error.

```cs
try
{
    return await url
        .AllowHttpStatus("100-299,6xx") // optional
        .GetJsonAsync<T>();
}
catch (FlurlHttpException ex)
{
   throw new Exception($"Server responded {ex.Call.Response.StatusCode}: {ex.GetResponseString()}");
}
```

### EasyHttp
[EasyHttp](https://github.com/EasyHttp/EasyHttp) throws an exception for any HTTP 4xx and 5xx response if you enable `client.ThrowExceptionOnHttpError` (disabled by default). The error-detection logic is internal, so can't be reused for custom handling.

```cs
HttpResponse response = client.Get(…);
if ((int)response.StatusCode >= 400 && (int)response.StatusCode <= 599)
   throw new Exception($"Server responded {response.StatusCode}: {response.RawText}");
return response.StaticBody<T>();
```

### FluentHttpClient (current)
For reference, this client throws exceptions on error.
```cs
try
{
    return await client.GetAsync(…).As<T>();
}
catch (ApiException ex)
{
   throw new Exception($"Server responded {ex.StatusCode}: {await ex.Response.AsString()}");
}
```

**Comment by Pathoschild** [comment]
### Use cases
I think FluentHttpClient aims to solve two main use cases:

* **Simple content fetch:** execute the request, parse the response, and only continue if it worked. For example:

  ```cs
  async Task<bool> IsUpdateAvailable(SemanticVersion current)
  {
     SemanticVersion latest = await new FluentHttpClient(…)
         .GetAsync("/releases/latest")
        .As<SemanticVersion>();
     return latest.IsNewerThan(current);
  }
  ```

  This is the default behaviour.

* **Fetch response as-is:** execute the request and read the response data as-is. If the request fails, your code is responsible for handling that. For example:

  ```cs
  async Task<bool> IsUpdateAvailable(SemanticVersion current)
  {
     IClient client = new FluentHttpClient();
     client.Filters.Remove<DefaultErrorFilter>();

     IResponse response = await client.GetAsync("/releases/latest");
     if (response.StatusCode == HttpStatusCode.OK)
        return (await response.As<SemanticVersion>()).IsNewerThan(current);
     return false;
  }
  ```

  This one works, but removing the `DefaultErrorFilter` is not very discoverable.

### Proposed solution
Here's one approach which makes both use cases intuitive:

1. Add a `ThrowIfFailed` method to `IHttpFilter` which throws an exception if the response is deemed erroneous, and move the `DefaultErrorFilter` logic to that.

   ```cs
   /// <summary>A middleware class which intercepts HTTP requests and responses. This can be used to implement common authentication, error-handling, etc.</summary>
   public interface IHttpFilter
   {
       /// <summary>Modify an outgoing HTTP request before it's submitted.</summary>
       /// <param name="request">The HTTP request.</param>
       void OnRequest(IRequest request);

       /// <summary>Modify an incoming HTTP response before it's parsed.</summary>
       /// <param name="response">The HTTP response.</param>
       void OnResponse(IResponse response);

       /// <summary>Throw an exception if the HTTP response failed according to this filter. This method isn't invoked when exceptions are suppressed.</summary>
       /// <param name="response">The HTTP response.</param>
       void ThrowIfFailed(IResponse response);
   }
   ```

2. The default behaviour is still to throw an exception when the request fails.

3. Add an argument to suppress exceptions on `IRequest` methods (by not calling `ThrowIfFailed`):

   ```cs
   IResponse response = await client.GetAsync(…, throwError: false);
   ```

This is more discoverable, since `throwOnError` will show up when you autocomplete the method (with a default value of `true`).

@Jericho: what do you think?

**Comment by Jericho** [comment]
I definitely like the discoverable `ThrowErrors` but I'm not convinced we need the 'ThrowIfFailed' method. Why not simply:
```CSharp
public void OnResponse(IResponse response, HttpResponseMessage responseMessage)
{
    if (!responseMessage.IsSuccessStatusCode && response.ThrowErrors)    // <-- that's probably not how to read the ThrowErrors property, but you get the idea...
        throw new ApiException(response, $"The API query failed with status code {responseMessage.StatusCode}: {responseMessage.ReasonPhrase}");
}
```
This would save developers from implementing the `ThrowIfFailed` method even if they don't need to when writing a custom HttpFilter.

**Comment by Pathoschild** [comment]
That would work, but I think we should distinguish between "real" exceptions (which might happen in `OnResponse`) and response-status-as-exceptions (which are an optional feature) to make the convention more discoverable when implementing a filter.

How about adding a base class so each implementation only overrides the methods it needs? For example:

```c#
/// <summary>An HTTP filter which detects failed HTTP requests and throws an exception.</summary>
public class DefaultErrorFilter : BaseHttpFilter
{
    /// <summary>Throw an exception if the HTTP response failed according to this filter. This method isn't invoked when exceptions are suppressed.</summary>
    /// <param name="response">The HTTP response.</param>
    public override void ThrowIfFailed(IResponse response)
    {
        if (!response.Message.IsSuccessStatusCode)
            throw new ApiException(response, $"The API query failed with status code {response.Message.StatusCode}: {response.Message.ReasonPhrase}");
    }
}
```

**Comment by Jericho** [comment]
If I have registered 4 filters, does that mean that 4 (potentially) different `ThrowIfFailed` would be invoked?

I would prefer if there was a way to designate a "error" filter which would be separate from all other filters. Maybe something like `client.SetErrorHandler(new MyCustomHandler());`

**Comment by Pathoschild** [comment]
Yep. The client will stop at the first one that throws an exception. That means it's easy to create handlers for specific error types:
```cs
client.Filters.Insert(0, new Error404Filter()); // throws KeyNotFoundException with a meaningful message
client.Filters.Insert(0, new Error418Filter()); // throws TeapotException
client.Filters.Insert(0, new Error500Filter()); // throws ApiException with server error details
```

**Comment by Jericho** [comment]
Any way to differentiate a error handling filter from other filters?

Maybe something like 
```CSharp
public class Error404Filter : IErrorHandler
```

**Comment by Pathoschild** [comment]
Maybe? I'd prefer to minimise the number of extensibility points though, to make it easier to document and understand. With my scheme, you only need to implement the methods you need but all filters are added the same way. What's the advantage of differentiating error filters?

**Comment by Jericho** [comment]
I am just trying to figure out a way where I don't need to worry about error handling when writing a "regular" filter and I don't need to worry about handling "regular" events when writing an error handler.

Besides, it would help with something similar to:
```CSharp
if (...an error has occurred...)
{
    foreach (var handler in this.Filters.OfType<IErrorHandler>())
    {
        handler.ThrowIfFailed(response);
    }
}
```

**Comment by Jericho** [comment]
Another thought just occurred to me: consider the scenario where I set `throwErrors` to false and also register `Error404Filter`. If there's no way to differentiate an error handler from a "regular" filter, we won't be able to respect "ThrowErrors".

**Comment by Jericho** [comment]
If you prefer to minimize extension points (which is perfectly legitimate objective), I would stick to the original filter interface (I don't see a need for `ThrowIfFailed`) and document the fact that each filter is responsible for checking the value of `ThrowErrors` and decide whether they want to respect it or not, similar to the [example](https://github.com/Pathoschild/FluentHttpClient/issues/38#issuecomment-274860523) I gave a little bit earlier.

**Comment by Pathoschild** [comment]
When writing a filter, you'd only need to worry about the methods you choose to override. The base class would have default empty implementations of all methods. It's the same approach used by [Web API action filters](https://msdn.microsoft.com/en-us/library/gg416513(vs.98).aspx), for example.

One thing to consider is that there isn't really such a thing as a "regular" or "error" filter — a filter is just a middleware class that intercepts requests and responses. For example, it's perfectly valid to create one filter that corrects all the weird issues for a given API (e.g. `FixAcmeApiFilter`), including both response munging and exception throwing.

The `if (...an error has occurred...)` part can't be handled by the client itself, because that's up to the filters to decide. The `DefaultErrorFilter` checks for the standard HTTP status codes, but supporting weird APIs (e.g. an API that returns errors as HTTP 200) is one of the main reasons for configurable filters.

The `ThrowIfFailed` method will only be called if `throwErrors` is `true`. That's why it's separate, and it's explicitly documented:
```c#
/// <summary>Throw an exception if the HTTP response failed according to this filter. This method isn't invoked when exceptions are suppressed.</summary>
/// <param name="response">The HTTP response.</param>
void ThrowIfFailed(IResponse response);
```

**Comment by Pathoschild** [comment]
Though maybe we could pass an argument like `bool throwErrorAsException` into `OnResponse`, instead of having a separate `ThrowIfFailed` method?

**Comment by Jericho** [comment]
Yes I prefer passing a `bool` to `OnResponse` and sticking to the original `IHttpFilter` interface . Or, even better: why not add a read-only `ThrowIfFailed` bool property on `IResponse`.

**Comment by Pathoschild** [comment]
It doesn't make sense as part of `IResponse`, because any exception has already been thrown by the time any code (except the filters themselves) get access to it. I'd be okay with a `bool` argument to `OnResponse`, though.

**Comment by Jericho** [comment]
I'm fine with that.

**Comment by Pathoschild** [comment]
So to recap, here's the implementation:
1. For `IRequest` methods which start the chain (like `GetAsync`), add a `bool httpErrorAsException = true` argument.
2. For `IHttpFilter::OnResponse`, add a `bool httpErrorAsException` argument.
2. Change `DefaultErrorFilter` to only throw if that argument is `true`.

Is that correct? Let me know if you want to tweak the argument names.

**Comment by Jericho** [comment]
I agree with 1, 2 and 3. Also, we need either a parameter in the FluentHttpClient constructor or a fluent method to set HttpErrorAsException.

Also, do you think we should pluralize: `HttpErrorsAsExceptions`?

**Comment by Pathoschild** [comment]
That's a good idea, but in that case the `IRequest` arguments would need to be something like `bool? httpErrorAsException = null` which is less intuitive.

**Comment by Jericho** [comment]
Why make it nullable? Let's just make a decision that `httpErrorAsException` is true by default (to match current behavior). I other words: `bool httpErrorAsException = true`

**Comment by Pathoschild** [comment]
That's fine by itself, but consider this code:

```c#
await new FluentClient(…)
   .SetHttpErrorAsException(false)
   .GetAsync("/");
```

Intuitively this shouldn't throw an exception, because you `SetHttpErrorAsException(false)` and didn't specify otherwise. But because `GetAsync` has a default `httpErrorAsException = true` value, that code is equivalent to this:

```c#
await new FluentClient(…)
   .SetHttpErrorAsException(false)
   .GetAsync("/", httpErrorAsException: true);
```

**Comment by Jericho** [comment]
I didn't realize you intended to allow setting `httpErrorAsException` at the request level in addition to the client level but I don't see that as a problem. In fact, this would be very similar to "Authentication" that we allow you to set the default at the client level and allow you to override it at the request level. We have `SetAuthentication` at the client level and `WithAuthentication` at the request level therefore, why not `SetHttpErrorAsException` at the client level and `WithHttpErrorAsException` at the request level.

Having said that (and at the risk of restarting this whole discussion from scratch, sorry about that), what if:
- we drop the idea of adding a `bool` parameter to the `OnResponse` method
- when an error filter is registered, it is always executed. Period.
- if you don't want an error handler to filter responses, just don't register it
- the `bool` at the client level simply causes the DefaultErrorFilter to be registered when set to true
- if the `bool` is set to false, the DefaultErrorFilter is simply not registered

This would allow me to set the bool to false to disable the default error handling and then to register my own custom error filter.

**Comment by Pathoschild** [comment]
I like the `GetAsync` etc argument because it's more discoverable, but I'd be okay with `Set/WithHttpErrorAsException(bool shouldThrow = true)` instead.

Having fluent methods toggle the `DefaultErrorFilter` would be confusing when you have custom filters in play. For example, the behaviour would not be intuitive in this case:

```c#
IClient client = new FluentHttpClient(…);
client.Filters.Add(new Error500Filter());

IResponse response = await client
    .SetHttpErrorAsException(false)
    .GetAsync(); // throws an exception on error 500
```

**Comment by Jericho** [comment]
To get rid of all this possible confusion, what do you think of the idea I [presented](https://github.com/Pathoschild/FluentHttpClient/issues/38#issuecomment-274917668) which would look like this:
```CSharp
IClient client = new FluentHttpClient(…);
client.DisableDefaultErrorHandler();
client.Filters.Add(new Error500Filter());

IResponse response = await client
    .GetAsync(); // custom error handler will throw an exception, default error handler is not registered

**Comment by Pathoschild** [comment]
That would work, but the fluent client is often used to build a client library for a particular API. If the library adds its own error filters, it'd be convenient as a user to be able to handle the response anyway:

```c#
IResponse response = await new AcmeApiClient(apiKey)
   .WithHttpErrorAsException(false) // suppress AcmeApiClient's custom http-error-as-exceptions too
   .GetAsync(…);
```

I'd be fine adding a `client.Filters.RemoveDefaultErrorHandling()` extension too though, since `DefaultErrorFilter` is an implementation detail.

**Comment by Jericho** [comment]
I'm not sure I understand your statement 
>it'd be convenient as a user to be able to handle the response anyway

I think what I'm suggesting allows exactly that: the developer can decide if the default error handler is registered or not and also add their own error handler if desired. They can have both the default handler AND custom filter registered at the same time if desired. I think that's the most flexible solution. 

To be clear, what I'm saying is:

- we drop the idea of adding a bool parameter to the OnResponse method
- when an error filter is registered, it is always executed. Period.
- if you don't want an error handler to filter responses, just don't register it
- by default, the default error handler is registered
- a fluent method at the client level allowing to remove the default error handler
- this fluent method can be called RemoveDefaultErrorHandler as you suggested.
- I think it's important that this fluent method be at the client level rather than the 'Filters' level because it would be much more discoverable. To be honest with you, it wasn't obvious to me when I started using FluentHttpClient that error handling was managed with `Filters` therefore it wouldn't occur to me to look under Filters for a way to remove the default handler.

**Comment by Pathoschild** [comment]
I see three use cases relevant to this discussion:

* **Scenario 1: one client for one request.**  
  This is the simplest case, with a one-off instance of the client. Any changes to the client only affect this request. No problem here.

  ```c#
  // ✓ proposal A
  IResponse response = await new FluentHttpClient(…)
     .DisableDefaultErrorHandler()
     .GetAsync(…); // OK!

  // ✓ proposal B
  IResponse response = await new FluentHttpClient(…)
     .WithHttpErrorAsException(false)
     .GetAsync(…); // OK!

  // ✓ proposal C
  IResponse response = await new FluentHttpClient(…)
     .GetAsync(…, httpErrorAsException: false); // OK!
  ```

* **Scenario 2: one client for multiple requests.**  
  This is a common case: set up the API client once, then send multiple requests with it. This is actually the [officially recommended approach](https://msdn.microsoft.com/en-us/library/system.net.http.httpclient(v=vs.110).aspx#Anchor_5). You can't change the client, because that would affect the other requests too.

  That causes a problem in proposal A: you can't change the client, but the only way to disable HTTP-errors-as-exceptions is to remove the `DefaultErrorFilter` on the client.

  ```c#
  // ✘ proposal A
  IResponse response = this.SharedClient
     //.DisableDefaultErrorHandler() // not allowed
     .GetAsync(…); // exception thrown

  // ✓ proposal B
  IResponse response = this.SharedClient
     .WithHttpErrorAsException(false)
     .GetAsync(…); // OK!

  // ✓ proposal C
  IResponse response = this.SharedClient
     .GetAsync(…, httpErrorAsException: false); // OK!
  ```

* **Scenario 3: use a wrapper library.**  
  In this case you aren't using `FluentHttpClient` directly. Instead you're using a wrapper client for a particular API which is preconfigured for the peculiarities of that API (with custom filters, request coordinator, default headers, etc).

  That causes two problems for proposal A: \[a\] since the client is preconfigured outside your control, 'just don't register the filters' doesn't apply; and \[b\] since there's no way to communicate your intention to the custom filters, they don't know they shouldn't throw HTTP errors as exceptions.

  ```c#
  // ✘ proposal A
  IResponse response = new AcmeClient(apiKey)
     .DisableDefaultErrorHandler()
     .GetAsync(…); // exception thrown by a custom filter

  // ✓ proposal B
  IResponse response = new AcmeClient(apiKey)
     .WithHttpErrorAsException(false)
     .GetAsync(…); // OK!

  // ✓ proposal C
  IResponse response = new AcmeClient(apiKey)
     .GetAsync(…, httpErrorAsException: false); // OK!
  ```

So with those scenarios, here's how the three current proposals stack up:

proposal | client for request | reusable client | wrapper library
-------- | -------------- | --------------- | ---------------
`bool httpErrorAsException` | ✓ OK! | ✓ OK! | ✓ OK!
`Set/WithHttpErrorAsException` | ✓ OK! | ✓ OK! | ✓ OK!
`DisableDefaultErrorHandler` | ✓ OK! | ✘ can't change client | ✘ internal filters

**Comment by Jericho** [comment]
_Note: the following paragraph is written in a 'tongue in check' tone, but I'm still trying to  make a serious point ;-)_

I feel like you just move the goal post! You added a new requirement: you don't want the client to be modified because that could affect other requests when a single client is shared. If that's the case, we need to remove all other fluent methods that also modify the client such as `SetAuthentication`, `SetUserAgent` and `SetRequestCoordinator`. There are also the public properties on the FluentHtpClient such as `BaseClient`, `Formatters` and `Filters`; they also impact other requests, are you thinking about removing them from the client as well? I'm being facetious, I know you don't intend to remove all these methods and public properties, but seriously I don't see what is so different about this one method that makes you uncomfortable but yet you are ok with these other methods and properties.

What I'm envisioning is that the method would simply look something like this:
```CSharp
public IClient RemoveDefaultErrorFilter()
{
    this.Filters.Remove<DefaultErrorFilter>();
    return this;
}
```

So, that makes me say that scenario 2 is not a deal breaker.

And regarding scenario 3: overriding the RemoveDefaultErrorFilter in the `Acme` client should take care of your concern. I envision something like this: 
```CSharp
public override IClient RemoveDefaultErrorFilter()
{
    base.RemoveDefaultErrorFilter()
    this.Filters.Remove<AcmeErrorFilter>();
    return this;
}
```

Also, I want to point out that your proposal C forces me to specify the Boolean value with every single Get, Post, Put, etc. which I find tedious. It's handy to have a way to override a setting when making one specific call (for example we have `WithAuthentication` which allows overriding the credentials for a specific request), but it's really convenient to have a global default value (we have `SetAuthentication` to set the default credentials).

At the end of the day, maybe the simplest thing to do is to keep the current error handling logic? Maybe developers read ReadMe more carefully than I did and they will notice the section when error handling is clearly explained.

I hope you don't feel I'm just being argumentative for the sake of it (I hate it when people do that to me!). I'm enjoying the back-and-forth discussion and I'm sure we'll figure out the best solution.

**Comment by Pathoschild** [comment]
Debate is healthy. I made some design mistakes early on because there was nobody to poke holes in them, so your argumentativeness is appreciated. ;)

No goalposts were moved — every client method (like `SetAuthentication`) has an equivalent request method (like `WithAuthentication`) so all three use cases are supported. Ideally everything that's configurable globally (`IClient::SetX`) should also be configurable per-request (`IRequest::WithX`), with only three exceptions so far:

setting | justification for being global-only
:------ | :-------------
`HttpClient` overrides | it would be too complicated to configure them per-request and there's little benefit to doing so.
Formatters | content negotiation makes per-request formatters unnecessary.
Filters | Maybe we should add `IRequest::WithFilter` and `IRequest::WithoutFilter`, actually. Might be a niche use case, though.

I agree about proposal C being tedious in some cases. How about merging proposals A and B?

1. Add `IClient::SetDefaultErrorFilter(false)`.  
   _This removes the default error-handling logic without caring about implementation details. Mostly useful when adding custom error filters._
2. Add `IClient::SetHttpErrorAsException(false)` and `IRequest::WithHttpErrorAsException(false)`.  
   _This would notify all filters (including custom filters) not to treat HTTP errors as exceptions, which addresses scenarios 2 & 3 above._
3. Change `DefaultErrorFilter` to only throw if `httpErrorAsException` is `true`.

**Comment by Pathoschild** [comment]
PS: maybe we can draft a _design principles_ document sometime to cover the conventions (like `SetX` vs `WithX`, using fluent methods, etc) and intended use cases (like the scenarios above)? It might be a useful reference when thinking of how we'd implement future features.

**Comment by Jericho** [comment]
1. `IClient::SetDefaultErrorFilter(true|false)` is perfect. It allows us to set at the client level whether we want the default error handler of not. I would even add `IRequest::WithDefaultErrorFilter(true|false)`.
2. Personally, I don't envision myself registering error filters and later on using `IClient::SetHttpErrorAsException(false)` and `IRequest::WithHttpErrorAsException(false)`. To me, that's just weird: the error filters are registered but effectively, they are turned off. I'm not opposed, but it wouldn't be useful to me. Also, there would be a performance penalty to loop through the error filters with `httpErrorAsException` set to `false` (which effectively would be a noop).
3. Indeed (but only if you decide to move forward with number 2).

**Comment by Pathoschild** [comment]
I do think `Set/WithHttpErrorAsException` would be useful in some scenarios as outlined above. It lets the user decide whether to use automatic error-handling in a semantic way, so custom filters can respect that intention. I think the impact of a few noop calls after a network await is negligible, but we could have a separate `OnResponseValidating` method that isn't called if `httpErrorAsException` is false.

**Comment by Pathoschild** [comment]
Can we move the `client.SetDefaultErrorFilter(bool)` method to `client.Filters.UseDefaultErrorFilter(bool)`? In retrospect...

* Having it on the root interface is confusing because:
  * When you're adding custom filters, you're accessing `client.Filters` anyway and would expect filter-related methods to be there.
  * When you aren't adding custom filters, it has the same effect as `client.SetHttpErrorsAsException(true|false)` but is less explicit (what _is_ the default error filter?).
* I think we should minimise the number of methods on the main fluent path, and I don't think configuring filters is part of the typical request flows.
* This sets a new convention for filter methods (e.g. a client library might have a `client.Filters.UsePhpArrayArgumentRewriter()` extension).

**Comment by Jericho** [comment]
I agree with changing the name to `UseDefault ...` rather than `SetDefault ...` but I do not agree with moving it from the root to `Filters`. The reason is because, when I started using FluentHtppClint, I didn't know that error handling was controlled by a filter. Therefore it wouldn't have occurred to me to look under `client.Filters` to find a way to turn on/off the default error handling.

Since this extension method seems to be so "controversial" maybe we should just forget about it and make sure developers are educated on how errors are handled and how they can remove the default handler by themselves. It's already on the ReadMe (although I didn't noticed it until you pointed it out to me). Maybe we need to make it bolder, bigger, more prominent, etc. and make sure developers don't overlook it like I did.

**Comment by Jericho** [comment]
And also, what about the idea of completely abandoning error handling? Developers have to either check the `StatusCode` (which they are used to doing with .net's HttpClient) or they write their own filter. Period. That's simpler for us and easy for them to understand!

**Comment by Pathoschild** [comment]
I don't think abandoning error-handling outright makes sense. I think ['simple content fetch'](https://github.com/Pathoschild/FluentHttpClient/issues/38#issuecomment-274673806) is the main use case, and there's no way for the user to check the response along those paths:

```c#
// how can the user check for a response code?
SomeModel model = await client.GetAsync(…).WithArguments(…).As<SomeModel>();
```

The `Set/WithHttpErrorsAsException` methods solve the problem where you can't check the response without dealing with filters, which are [an advanced feature](https://github.com/Pathoschild/FluentHttpClient/tree/2.3#customising-the-client) that most users shouldn't need to know about. Those methods are on the root interfaces and they're discoverable for new users (since they'll appear when you type `.error`).

I'm okay removing `SetDefaultErrorFilter` since you can do the same thing with `filters.Remove<DefaultErrorFilter>`, but it'd be nice to have a `filter.RemoveDefaultErrorFilter()` so you don't need to know about the `DefaultErrorFilter` class which isn't even in the main namespace.

**Comment by Pathoschild** [comment]
Added `Set/WithHttpErrorAsException` in `develop` for the upcoming 3.0 release; pending a decision on the `SetDefaultErrorFilter` method.

**Comment by Jericho** [comment]
The fact that I can add/remove error filter on the client and also on the request (when PR #43 is merged), I am completely happy and I fee all my scenarios are addressed. I can remove the default handler if I want to, I can add my own handler and I can remove the handler from a specific request. As far as I'm concerned, this addresses all my requirements. I am now well educated on how this works and I find it very easy.

**Comment by Pathoschild** [comment]
Excellent. The changes for this ticket are done and #43 is merged into the 3.0 release, so I'll mark this done.
